---
title: "Script_1_4_General_Sample_Overview"
author: Linda Dieckmann
date: November 2022
output: github_document
---

In this notebook we take a look at the number and overlap of the samples and make a descriptive table.

# load packages
```{r}
library(here)
library(RColorBrewer)
library(data.table)
library(VennDiagram)
library(plyr)
library(gtsummary)
```

If you update package(s) call 
  renv::snapshot() 
to save the current picture of the project.

```{r}
writeLines(capture.output(sessionInfo()), here("Analyses/Scripts/sessionInfos/", "sessionInfo_Script_1_4.txt"))
```

If you update package(s) call renv::snapshot() to save the current picture of the project.
```{r}
writeLines(capture.output(sessionInfo()), here("03_Code_Scripts/", "sessionInfo_Script_1.4.txt"))
```


```{r}
source(here("Analyses/Scripts/", "functions.R"))
```

# load data
```{r}
load(here("02_Data/raw/", "geno_t_fullqced_IDc.Rdata"))
load(here("02_Data/prepared/", "inv_norm_placenta_tmm_filtered.Rdata"))
load(here("02_Data/prepared/", "inv_norm_cvs_tmm_filtered.Rdata"))
load(here("02_Data/prepared/", "methylation_M_placenta_filtered.Rdata"))
load(here("02_Data/prepared/", "methylation_M_cvs_filtered.Rdata"))
load(here("02_Data/prepared", "common_eqtl_placenta_ids.Rdata"))
load(here("02_Data/prepared", "common_eqtl_cvs_ids.Rdata"))
load(here("02_Data/prepared", "common_meqtl_placenta_ids.Rdata"))
load(here("02_Data/prepared", "common_meqtl_cvs_ids.Rdata"))
load(here("02_Data/prepared", "common_eqtm_cvs_ids.Rdata"))
load(here("02_Data/prepared", "common_eqtm_placenta_ids.Rdata"))
load(here("02_Data/prepared", "cov_meth_placenta.Rdata"))
load(here("02_Data/prepared", "cov_meth_cvs.Rdata"))
load(here("02_Data/prepared", "cov_rna_placenta.Rdata"))
load(here("02_Data/prepared", "cov_rna_cvs.Rdata"))
```

```{r}
load(here("02_Data/raw/", "ITU_register_vars.Rdata"))
```

# Sample Overview
```{r}
myCol <- brewer.pal(3, "Pastel2")
```

## Placenta biological data
```{r}
futile.logger::flog.threshold(futile.logger::ERROR, name = "VennDiagramLogger")
venn.diagram(
  x = list(colnames(geno_t_fullqced_IDc), colnames(methylation_M_placenta_filtered), colnames(inv_norm_placenta_tmm_filtered)),
  category.names = c("Genotype" , "DNAm" , "RNA-Seq"),
  filename = here("Analyses/Results/Descriptive/", "placenta_samples_venn.png"),
  output=TRUE,
  
  # Output features
  imagetype="png" ,
  height = 600 , 
  width = 600 , 
  resolution = 300,
  compression = "lzw",
  
  # Circles
  lwd = 2,
  lty = 'blank',
  fill = c(colors_genotypes, color_placenta_methylation, color_placenta_rna),
  
  # Numbers
  cex = .8,
  fontface = "bold",
  fontfamily = "sans",
  
  # Set names
  cat.cex = 0.6,
  cat.fontface = "bold",
  cat.default.pos = "outer",
  cat.pos = c(-27, 27, 135),
  cat.dist = c(0.055, 0.055, 0.085),
  cat.fontfamily = "sans",
  rotation = 1
)

```

## CVS biological data
```{r}
futile.logger::flog.threshold(futile.logger::ERROR, name = "VennDiagramLogger")
venn.diagram(
  x = list(colnames(geno_t_fullqced_IDc), colnames(methylation_M_cvs_filtered), colnames(inv_norm_cvs_tmm_filtered)),
  category.names = c("Genotype" , "DNAm" , "RNA-Seq"),
  filename = here("Analyses/Results/Descriptive/", "cvs_samples_venn.png"),
  output=TRUE,
  
  # Output features
  imagetype="png" ,
  height = 600 , 
  width = 600 , 
  resolution = 300,
  compression = "lzw",
  
  # Circles
  lwd = 2,
  lty = 'blank',
  fill = c(colors_genotypes, color_cvs_methylation, color_cvs_rna),
  
  # Numbers
  cex = .8,
  fontface = "bold",
  fontfamily = "sans",
  
  # Set names
  cat.cex = 0.6,
  cat.fontface = "bold",
  cat.default.pos = "outer",
  cat.pos = c(-27, 27, 135),
  cat.dist = c(0.055, 0.055, 0.085),
  cat.fontfamily = "sans",
  rotation = 1
)
```


# intersection of all 3 omics-levels
## placenta
```{r}
common_omics_placenta_ids <- Reduce(intersect, list(colnames(methylation_M_placenta_filtered), colnames(inv_norm_placenta_tmm_filtered), colnames(geno_t_fullqced_IDc), colnames(cov_meth_placenta), colnames(cov_rna_placenta)))
length(common_omics_placenta_ids)
```

## cvs
```{r}
common_omics_cvs_ids <- Reduce(intersect, list(colnames(methylation_M_cvs_filtered), colnames(inv_norm_cvs_tmm_filtered), colnames(geno_t_fullqced_IDc), colnames(cov_meth_cvs), colnames(cov_rna_cvs)))
length(common_omics_cvs_ids)
```


# Sample Overview QTL analyses - CVS & Placenta

## eQTLs
```{r}
futile.logger::flog.threshold(futile.logger::ERROR, name = "VennDiagramLogger")
venn.diagram(
  x = list(common_eqtl_placenta_ids, common_eqtl_cvs_ids),
  category.names = c("Placenta" , "CVS"),
  filename = here("Analyses/Results/Descriptive/", "eqtls-cvs-placenta_venn.png"),
  output=TRUE,
  
  # Output features
  imagetype="png" ,
  height = 600 , 
  width = 600 , 
  resolution = 300,
  compression = "lzw",
  
  # Circles
  lwd = 2,
  lty = 'blank',
  fill = colors_placenta_cvs_colored,
  
  # Numbers
  cex = .8,
  fontface = "bold",
  fontfamily = "sans",
  
  # Set names
  cat.cex = 0.6,
  cat.fontface = "bold",
  cat.default.pos = "outer",
  cat.fontfamily = "sans",
  cat.pos = c(0, 180),
  cat.dist = c(0.02, 0.02),
  rotation.degree = 180
)
```

## meQTLs
```{r}
futile.logger::flog.threshold(futile.logger::ERROR, name = "VennDiagramLogger")
venn.diagram(
  x = list(common_meqtl_placenta_ids, common_meqtl_cvs_ids),
  category.names = c("Placenta", "CVS"),
  filename = here("Analyses/Results/Descriptive/", "meqtls-cvs-placenta_venn.png"),
  output=TRUE,
  
  # Output features
  imagetype="png" ,
  height = 600 , 
  width = 600 , 
  resolution = 300,
  compression = "lzw",
  
  # Circles
  lwd = 2,
  lty = 'blank',
  fill = colors_placenta_cvs_colored,
  
  # Numbers
  cex = .8,
  fontface = "bold",
  fontfamily = "sans",
  
  # Set names
  cat.cex = 0.6,
  cat.fontface = "bold",
  cat.default.pos = "outer",
  cat.fontfamily = "sans",
  cat.pos = c(0, 180),
  cat.dist = c(0.02, 0.02),
  rotation.degree = 180
)
```

## eQTMs
```{r}
futile.logger::flog.threshold(futile.logger::ERROR, name = "VennDiagramLogger")
venn.diagram(
  x = list(common_eqtm_placenta_ids, common_eqtm_cvs_ids),
  category.names = c("Placenta","CVS"),
  filename = here("Analyses/Results/Descriptive/", "eqtms-cvs-placenta_venn.png"),
  output=TRUE,
  
  # Output features
  imagetype="png" ,
  height = 600 , 
  width = 600 , 
  resolution = 300,
  compression = "lzw",
  
  # Circles
  lwd = 2,
  lty = 'blank',
  fill = colors_placenta_cvs_colored,
  
  # Numbers
  cex = .8,
  fontface = "bold",
  fontfamily = "sans",
  
  # Set names
  cat.cex = 0.6,
  cat.fontface = "bold",
  cat.default.pos = "outer",
  cat.fontfamily = "sans",
  cat.pos = c(0, 180),
  cat.dist = c(0.02, 0.02),
  rotation.degree = 180
)
```

# Descriptive Table
All samples which were included in at least one of the analyses (eQTLs/meQTLs/eQTMs) for placenta vs. CVS.

```{r}
placenta_ids <- unique(c(common_eqtl_placenta_ids, common_meqtl_placenta_ids, common_eqtm_placenta_ids))
length(placenta_ids)
cvs_ids <- unique(c(common_eqtl_cvs_ids, common_meqtl_cvs_ids, common_eqtm_cvs_ids))
length(cvs_ids)
all_ids <- unique(c(placenta_ids, cvs_ids))
length(all_ids)
```

```{r}
ITU_phenos_cvs <- ITU_register_vars[ITU_register_vars$IDfamily %in% cvs_ids, ]
ITU_phenos_cvs$Child_Birth_Weight_kg <- ITU_phenos_cvs$Child_Birth_Weight / 1000
ITU_phenos_cvs$Placental_Weight_kg <- ITU_phenos_cvs$Placental_Weight_Grams / 1000

ITU_phenos_placenta <- ITU_register_vars[ITU_register_vars$IDfamily %in% placenta_ids, ]
ITU_phenos_placenta$Child_Birth_Weight_kg <- ITU_phenos_placenta$Child_Birth_Weight / 1000
ITU_phenos_placenta$Placental_Weight_kg <- ITU_phenos_placenta$Placental_Weight_Grams / 1000
```

```{r}
ITU_phenos_cvs_desc <- ITU_phenos_cvs[ ,c("Child_Birth_Weight_kg", "Child_Birth_Length", "Placental_Weight_kg", "Gestational_Age_Weeks", "gestage_at_CVS_weeks", "Child_Sex", "Maternal_Age_Years", "smoking_dichotom", "Maternal_Diabetes_dichotom", "Delivery_mode_3l")]

ITU_phenos_cvs_desc$Maternal_Diabetes_dichotom <- mapvalues(ITU_phenos_cvs_desc$Maternal_Diabetes_dichotom, from = c("no diabetes this pregnancy", "diabetes in current pregnancy"), to = c("not in current pregnancy", "yes, in current pregnancy"))
ITU_phenos_cvs_desc$Delivery_mode_3l <- mapvalues(ITU_phenos_cvs_desc$Delivery_mode_3l, from = c("Vaginal_all", "Planned caesarian section", "urgent caesarian section"), to = c("vaginal", "planned cesarian section", "urgent cesarian section"))
```

```{r}
ITU_phenos_placenta_desc <- ITU_phenos_placenta[ ,c("Child_Birth_Weight_kg", "Child_Birth_Length", "Placental_Weight_kg", "Gestational_Age_Weeks", "gestage_at_CVS_weeks", "Child_Sex", "Maternal_Age_Years", "smoking_dichotom", "Maternal_Diabetes_dichotom", "Delivery_mode_3l")]

ITU_phenos_placenta_desc$Maternal_Diabetes_dichotom <- mapvalues(ITU_phenos_placenta_desc$Maternal_Diabetes_dichotom, from = c("no diabetes this pregnancy", "diabetes in current pregnancy"), to = c("not in current pregnancy", "yes, in current pregnancy"))
ITU_phenos_placenta_desc$Delivery_mode_3l <- mapvalues(ITU_phenos_placenta_desc$Delivery_mode_3l, from = c("Vaginal_all", "Planned caesarian section", "urgent caesarian section"), to = c("vaginal", "planned cesarian section", "urgent cesarian section"))
```

```{r}
CVS_descriptive_summary_table <- 
  ITU_phenos_cvs_desc %>%
  tbl_summary(
    missing = "no", 
    label = list(Child_Birth_Weight_kg ~ "birth weight [kg]", Child_Birth_Length ~ "birth length [cm]", Placental_Weight_kg ~ "placental weight [kg]", Gestational_Age_Weeks ~ "gestational age at birth [weeks]", gestage_at_CVS_weeks ~ "gestational age at CVS sampling [weeks]", Child_Sex ~ "sex", Maternal_Age_Years ~ "maternal age [years]", smoking_dichotom ~ "maternal smoking", Maternal_Diabetes_dichotom ~ "maternal diabetes in current pregnancy", Delivery_mode_3l ~ "delivery mode"),
    type = list(c(Child_Birth_Weight_kg, Child_Birth_Length, Placental_Weight_kg, Gestational_Age_Weeks, gestage_at_CVS_weeks, Maternal_Age_Years) ~ "continuous", c(Child_Sex, smoking_dichotom, Maternal_Diabetes_dichotom) ~ "dichotomous"),
    value = list(Child_Sex ~ "male", smoking_dichotom ~ "yes", Maternal_Diabetes_dichotom ~ "yes, in current pregnancy"),
    statistic = all_continuous() ~"{mean} ({sd})",
    digits = all_continuous() ~ 2
  ) %>%
  add_n(last = T)  %>%
  modify_header(label ~ "*Phenotype*", n ~ "n", stat_0 ~ "**CVS**")
```

```{r}
Placenta_descriptive_summary_table <- 
  ITU_phenos_placenta_desc %>%
  tbl_summary(
    missing = "no", 
    label = list(Child_Birth_Weight_kg ~ "birth weight [kg]", Child_Birth_Length ~ "birth length [cm]", Placental_Weight_kg ~ "placental weight [kg]", Gestational_Age_Weeks ~ "gestational age at birth [weeks]", gestage_at_CVS_weeks ~ "gestational age at CVS sampling [weeks]", Child_Sex ~ "sex", Maternal_Age_Years ~ "maternal age [years]", smoking_dichotom ~ "maternal smoking", Maternal_Diabetes_dichotom ~ "maternal diabetes in current pregnancy", Delivery_mode_3l ~ "delivery mode"),
    type = list(c(Child_Birth_Weight_kg, Child_Birth_Length, Placental_Weight_kg, Gestational_Age_Weeks, gestage_at_CVS_weeks, Maternal_Age_Years) ~ "continuous", c(Child_Sex, smoking_dichotom, Maternal_Diabetes_dichotom) ~ "dichotomous"),
    value = list(Child_Sex ~ "male", smoking_dichotom ~ "yes", Maternal_Diabetes_dichotom ~ "yes, in current pregnancy"),
    statistic = all_continuous() ~"{mean} ({sd})",
    digits = all_continuous() ~ 2
  ) %>%
  add_n(last = T)  %>%
  modify_header(label ~ "*Phenotype*", n ~ "n", stat_0 ~ "**Placenta**")
```


```{r}
CVS_descriptive_summary_table
Placenta_descriptive_summary_table
```

```{r}
desc_table <- tbl_merge(list(CVS_descriptive_summary_table, Placenta_descriptive_summary_table), tab_spanner = F)
gt_desc_table <- as_gt(desc_table)
```

```{r}
gt_desc_table %>%
  gt::gtsave(filename = here::here("Analyses/Results/Descriptive/", "Sample_Descriptive_Table.docx"))
```





