---
title: "Basic Results meQTLs from MatrixEQTL and BootstrapQTL, common probes & samples, scaled"
author: Linda Dieckmann
date: January 2023
output: github_document
---

In this notebook we take take a look at basic results from MatrixEQTL vs. BootstrapQTL for meQTLs for the samples in CVS + Placenta

# load packages
```{r, warning=FALSE}
library(here)
library(MatrixEQTL)
library(BootstrapQTL)
library(ggplot2)
library(dplyr)
library(cowplot)
library(scales)
library(data.table)
library(rcompanion)
```

If you update package(s) call 
  renv::snapshot() 
to save the current picture of the project.

```{r}
writeLines(capture.output(sessionInfo()), here("Analyses/Scripts/sessionInfos/", "sessionInfo_Script_3_2_cp.txt"))
```

```{r}
source(here("Analyses/Scripts/", "functions.R"))
```

################################################################################

# load data

## QTL analyses results
```{r}
# meQTL

# MatrixEQTL
load(here("02_Data/MatrixEQTL_Output", "scaled_me_meqtl_placenta_cp_samples_probes.Rdata"))
load(here("02_Data/MatrixEQTL_Output", "scaled_me_meqtl_cvs_cp_samples_probes.Rdata"))

# Bootstrap QTL (data frame with results for significant hits)
load(here("02_Data/MatrixEQTL_Output", "scaled_bme_meqtl_placenta_cp_samples_probes.Rdata"))
load(here("02_Data/MatrixEQTL_Output", "scaled_bme_meqtl_cvs_cp_samples_probes.Rdata"))
```

sample names
```{r}
load(here("02_Data/prepared", "common_meqtl_samples_cvs_placenta.Rdata"))
```

gaphunter probes
```{r}
gaps_placenta <- read.table(here("02_Data/raw/gaphunter/final_data", "gaphunter_probes_placenta.txt"))
gaps_cvs <- read.table(here("02_Data/raw/gaphunter/final_data", "gaphunter_probes_cvs.txt"))
```

gene info
```{r}
load(here("02_Data/raw/", "geno_info.Rdata")) # info from bim file
```

```{r}
genotypes <- geno_info[ ,c("snp.name", "chromosome", "position", "allele.2")]
names(genotypes) <- c("snp.name", "snp.chromosome", "snp.position", "effect.allele")
```

## placenta

### MatrixEQTL
standard meqtls output (linear model (t-statistic), only cis associations)
```{r}
scaled_statistics_me_meqtl_placenta_cp_samples_probes_cis <- scaled_me_meqtl_placenta_cp_samples_probes$cis$eqtls
```

we add standard error and r2
```{r}
scaled_statistics_me_meqtl_placenta_cp_samples_probes_cis$se_beta <- scaled_statistics_me_meqtl_placenta_cp_samples_probes_cis$beta / scaled_statistics_me_meqtl_placenta_cp_samples_probes_cis$statistic
```

```{r}
scaled_df_meqtl_placenta_cp_samples_probes <- scaled_me_meqtl_placenta_cp_samples_probes$param$dfFull
scaled_r_meqtl_placenta_cp_samples_probes <- scaled_statistics_me_meqtl_placenta_cp_samples_probes_cis$statistic / sqrt(scaled_df_meqtl_placenta_cp_samples_probes + scaled_statistics_me_meqtl_placenta_cp_samples_probes_cis$statistic^2)
scaled_r2_meqtl_placenta_cp_samples_probes <- scaled_r_meqtl_placenta_cp_samples_probes ^ 2
```

```{r}
scaled_statistics_me_meqtl_placenta_cp_samples_probes_cis$r2 <- scaled_r2_meqtl_placenta_cp_samples_probes
```

How many tests were performed?
```{r}
scaled_me_meqtl_placenta_cp_samples_probes$cis$ntests
```

How many cis associations with FDR < 0.05 were found with MatrixEQTL?
```{r}
scaled_statistics_me_meqtl_placenta_cp_samples_probes_cis_fdr5 <- scaled_statistics_me_meqtl_placenta_cp_samples_probes_cis[scaled_statistics_me_meqtl_placenta_cp_samples_probes_cis$FDR < 0.05, ]
nrow(scaled_statistics_me_meqtl_placenta_cp_samples_probes_cis_fdr5)
```

add a column indicating if FDR < 0.05 to the total MatrixEQTL results
```{r}
scaled_statistics_me_meqtl_placenta_cp_samples_probes_cis$FDRsig <- ifelse(scaled_statistics_me_meqtl_placenta_cp_samples_probes_cis$FDR < 0.05, "yes", "no")
# yes if FDR < 0.05
```
 
How many unique SNPs with FDR < 0.05 were found with Matrixmeqtl?
```{r}
scaled_unique_snp_names_meqtl_placenta_fdr5_matrix <- unique(scaled_statistics_me_meqtl_placenta_cp_samples_probes_cis_fdr5$snps)
length(scaled_unique_snp_names_meqtl_placenta_fdr5_matrix)
```

How many unique cpgs with FDR < 0.05 were found with Matrixmeqtl?
```{r}
scaled_unique_cpg_names_meqtl_placenta_fdr5_matrix <- unique(scaled_statistics_me_meqtl_placenta_cp_samples_probes_cis_fdr5$gene)
length(scaled_unique_cpg_names_meqtl_placenta_fdr5_matrix)
```

### Bootstrap QTL
How many significant cis associations were found with Bootstrap QTL (global correction p < 0.05)?
```{r}
nrow(scaled_bme_meqtl_placenta_cp_samples_probes)
```

How many unique SNPs do we have?
```{r}
scaled_unique_snp_names_meqtl_placenta_boot <- unique(scaled_bme_meqtl_placenta_cp_samples_probes$eSNPs)
length(scaled_unique_snp_names_meqtl_placenta_boot)
```

How many unqiue cpgs do we have?
```{r}
scaled_unique_cpg_names_meqtl_placenta_boot <- unique(scaled_bme_meqtl_placenta_cp_samples_probes$eGene)
length(scaled_unique_cpg_names_meqtl_placenta_boot)
```

```{r}
scaled_bme_meqtl_placenta_cp_samples_probes$inboot<- "yes"
# if among the bootstrapqtl results, yes
```

*combine MatrixEQTL and BootstrapQTL Data*
```{r}
scaled_meQTLs_placenta_statistics_cp_samples_probes <- merge(scaled_bme_meqtl_placenta_cp_samples_probes[,-3], scaled_statistics_me_meqtl_placenta_cp_samples_probes_cis, all = T, by.x=c("eGene", "eSNPs"), by.y=c("gene", "snps"))
```

```{r}
# if na in inboot, it means it is not among bootstrap sig results
scaled_meQTLs_placenta_statistics_cp_samples_probes[c("inboot")][is.na(scaled_meQTLs_placenta_statistics_cp_samples_probes[c("inboot")])] <- "no"
```

add gaphunter probes
```{r}
scaled_meQTLs_placenta_statistics_cp_samples_probes$gap_probe <- ifelse(scaled_meQTLs_placenta_statistics_cp_samples_probes$eGene %in% gaps_placenta$V1, "yes", "no")
```

*add geno info*
```{r}
scaled_meQTLs_placenta_statistics_cp_samples_probes <- merge(scaled_meQTLs_placenta_statistics_cp_samples_probes, genotypes, by.x="eSNPs", by.y="snp.name")
```

```{r}
colnames(scaled_meQTLs_placenta_statistics_cp_samples_probes) = gsub("Gene", "CpG", colnames(scaled_meQTLs_placenta_statistics_cp_samples_probes))
```

### ODDs
```{r}
total_number_tests_placenta <- scaled_me_meqtl_placenta_cp_samples_probes$cis$ntests
number_positive_placenta <- nrow(scaled_bme_meqtl_placenta_cp_samples_probes)
number_negative_placenta <- total_number_tests_placenta - number_positive_placenta

odds_placenta <- number_positive_placenta / number_negative_placenta

cat("Odds of significant association in placenta are", number_positive_placenta, "/", number_negative_placenta)
```

### export full meQTL placenta statistic table
```{r, warning = F}
write.csv(scaled_meQTLs_placenta_statistics_cp_samples_probes, file = here("Analyses/Results/Descriptive", "scaled_meQTLs_placenta_statistics_cp_samples_probes.csv"), row.names = FALSE, col.names = T)
save(scaled_meQTLs_placenta_statistics_cp_samples_probes, file = here("02_Data/MatrixEQTL_Output", "scaled_meQTLs_placenta_statistics_cp_samples_probes.Rdata"))
```

### histogram plot

```{r}
scaled_hits_df_meqtls_placenta_cp_samples_probes <- data.frame(c(nrow(scaled_statistics_me_meqtl_placenta_cp_samples_probes_cis_fdr5), length(scaled_unique_snp_names_meqtl_placenta_fdr5_matrix), length(scaled_unique_cpg_names_meqtl_placenta_fdr5_matrix)), c(nrow(scaled_bme_meqtl_placenta_cp_samples_probes), length(scaled_unique_snp_names_meqtl_placenta_boot), length(scaled_unique_cpg_names_meqtl_placenta_boot)))

rownames(scaled_hits_df_meqtls_placenta_cp_samples_probes) <- c("cis meQTLs", "unique SNPs", "unique CpGs")
colnames(scaled_hits_df_meqtls_placenta_cp_samples_probes) <- c("MatrixEqtl", "BootstrapQTL")

scaled_hits_df_meqtls_placenta_cp_samples_probes <- scaled_hits_df_meqtls_placenta_cp_samples_probes %>% as.matrix() %>% reshape2::melt(varnames = c("number", "method"), value.name = "cnt")
```


```{r}
barplot_counts_function(scaled_hits_df_meqtls_placenta_cp_samples_probes, "Placenta", "BootstrapQTL", colors_placenta_3levels, ylimits=c(0,16200000))
```

##CVS

### MatrixEQTL
standard meqtls output (linear model (t-statistic), only cis associations)
```{r}
scaled_statistics_me_meqtl_cvs_cp_samples_probes_cis <- scaled_me_meqtl_cvs_cp_samples_probes$cis$eqtls
```

we add standard error and r2
```{r}
scaled_statistics_me_meqtl_cvs_cp_samples_probes_cis$se_beta <- scaled_statistics_me_meqtl_cvs_cp_samples_probes_cis$beta / scaled_statistics_me_meqtl_cvs_cp_samples_probes_cis$statistic
```

```{r}
scaled_df_meqtl_cvs <- scaled_me_meqtl_cvs_cp_samples_probes$param$dfFull
scaled_r_meqtl_cvs <- scaled_statistics_me_meqtl_cvs_cp_samples_probes_cis$statistic / sqrt(scaled_df_meqtl_cvs + scaled_statistics_me_meqtl_cvs_cp_samples_probes_cis$statistic^2)
scaled_r2_meqtl_cvs <- scaled_r_meqtl_cvs ^ 2
```

```{r}
scaled_statistics_me_meqtl_cvs_cp_samples_probes_cis$r2 <- scaled_r2_meqtl_cvs
```

How many tests were performed?
```{r}
scaled_me_meqtl_cvs_cp_samples_probes$cis$ntests
```

How many cis associations with FDR < 0.05 were found with MatrixEQTL?
```{r}
scaled_statistics_me_meqtl_cvs_cp_samples_probes_cis_fdr5 <- scaled_statistics_me_meqtl_cvs_cp_samples_probes_cis[scaled_statistics_me_meqtl_cvs_cp_samples_probes_cis$FDR < 0.05, ]
nrow(scaled_statistics_me_meqtl_cvs_cp_samples_probes_cis_fdr5)
```

add a column indicating if FDR < 0.05 to the total MatrixEQTL results
```{r}
scaled_statistics_me_meqtl_cvs_cp_samples_probes_cis$FDRsig <- ifelse(scaled_statistics_me_meqtl_cvs_cp_samples_probes_cis$FDR < 0.05, "yes", "no")
# yes if FDR < 0.05
```

How many unique SNPs with FDR < 0.05 were found with MatrixEQTL?
```{r}
scaled_unique_snp_names_meqtl_cvs_fdr5_matrix <- unique(scaled_statistics_me_meqtl_cvs_cp_samples_probes_cis_fdr5$snps)
length(scaled_unique_snp_names_meqtl_cvs_fdr5_matrix)
```

How many unique cpgs with FDR < 0.05 were found with MatrixEQTL?
```{r}
scaled_unique_cpg_names_meqtl_cvs_fdr5_matrix <- unique(scaled_statistics_me_meqtl_cvs_cp_samples_probes_cis_fdr5$gene)
length(scaled_unique_cpg_names_meqtl_cvs_fdr5_matrix)
```


### Bootstrap QTL
How many significant cis associations were found with Bootstrap QTL (global correction p < 0.05)?
```{r}
nrow(scaled_bme_meqtl_cvs_cp_samples_probes)
```

How many unique SNPs do we have?
```{r}
scaled_unique_snp_names_meqtl_cvs_boot <- unique(scaled_bme_meqtl_cvs_cp_samples_probes$eSNPs)
length(scaled_unique_snp_names_meqtl_cvs_boot)
```
How many unqiue cpgs do we have?
```{r}
scaled_unique_cpg_names_meqtl_cvs_boot <- unique(scaled_bme_meqtl_cvs_cp_samples_probes$eGene)
length(scaled_unique_cpg_names_meqtl_cvs_boot)
```

```{r}
scaled_bme_meqtl_cvs_cp_samples_probes$inboot<- "yes"
# if among the bootstrapqtl results, yes
```

*combine MatrixEQTL and BootstrapQTL Data*
```{r}
scaled_meQTLs_cvs_statistics_cp_samples_probes <- merge(scaled_bme_meqtl_cvs_cp_samples_probes[,-3], scaled_statistics_me_meqtl_cvs_cp_samples_probes_cis, all = T, by.x=c("eGene", "eSNPs"), by.y=c("gene", "snps"))
```

```{r}
scaled_meQTLs_cvs_statistics_cp_samples_probes[c("inboot")][is.na(scaled_meQTLs_cvs_statistics_cp_samples_probes[c("inboot")])] <- "no"
```

add gaphunter probes
```{r}
scaled_meQTLs_cvs_statistics_cp_samples_probes$gap_probe <- ifelse(scaled_meQTLs_cvs_statistics_cp_samples_probes$eGene %in% gaps_cvs$V1, "yes", "no")
```

*add geno info*
```{r}
scaled_meQTLs_cvs_statistics_cp_samples_probes <- merge(scaled_meQTLs_cvs_statistics_cp_samples_probes, genotypes, by.x="eSNPs", by.y="snp.name")
```

```{r}
colnames(scaled_meQTLs_cvs_statistics_cp_samples_probes) = gsub("Gene", "CpG", colnames(scaled_meQTLs_cvs_statistics_cp_samples_probes))
```

### ODDs
```{r}
total_number_tests_cvs <- scaled_me_meqtl_cvs_cp_samples_probes$cis$ntests
number_positive_cvs <- nrow(scaled_bme_meqtl_cvs_cp_samples_probes)
number_negative_cvs <- total_number_tests_cvs - number_positive_cvs

odds_cvs <- number_positive_cvs / number_negative_cvs

cat("Odds of significant association in cvs are", number_positive_cvs, "/", number_negative_cvs)
```

### export full eQTL placenta statistic table
```{r, warning=F}
write.csv(scaled_meQTLs_cvs_statistics_cp_samples_probes, file = here("Analyses/Results/Descriptive", "scaled_meQTLs_cvs_statistics_cp_samples_probes.csv"), row.names = FALSE, col.names = T)
save(scaled_meQTLs_cvs_statistics_cp_samples_probes, file = here("02_Data/MatrixEQTL_Output", "scaled_meQTLs_cvs_statistics_cp_samples_probes.Rdata"))
```

### histogram plot
```{r}
scaled_hits_df_meqtls_cvs_cp_samples_probes <- data.frame(c(nrow(scaled_statistics_me_meqtl_cvs_cp_samples_probes_cis_fdr5), length(scaled_unique_snp_names_meqtl_cvs_fdr5_matrix), length(scaled_unique_cpg_names_meqtl_cvs_fdr5_matrix)), c(nrow(scaled_bme_meqtl_cvs_cp_samples_probes), length(scaled_unique_snp_names_meqtl_cvs_boot), length(scaled_unique_cpg_names_meqtl_cvs_boot)))

rownames(scaled_hits_df_meqtls_cvs_cp_samples_probes) <- c("cis meQTLs", "unique SNPs", "unique CpGs")
colnames(scaled_hits_df_meqtls_cvs_cp_samples_probes) <- c("MatrixEqtl", "BootstrapQTL") 

scaled_hits_df_meqtls_cvs_cp_samples_probes <- scaled_hits_df_meqtls_cvs_cp_samples_probes %>% as.matrix() %>% reshape2::melt(varnames = c("number", "method"), value.name = "cnt")
```


```{r}
barplot_counts_function(scaled_hits_df_meqtls_cvs_cp_samples_probes, "cvs", "BootstrapQTL", colors_cvs_3levels, ylimits=c(0,16200000))
```

# combination CVS, Placenta
```{r}
scaled_meQTLs_statistics_cp_samples_probes_placenta_cvs <- merge(scaled_meQTLs_placenta_statistics_cp_samples_probes, scaled_meQTLs_cvs_statistics_cp_samples_probes, by = c("eSNPs", "eCpG"))
cat(nrow(scaled_meQTLs_statistics_cp_samples_probes_placenta_cvs))
```


```{r}
colnames(scaled_meQTLs_statistics_cp_samples_probes_placenta_cvs) <- gsub('.x', '.placenta', colnames(scaled_meQTLs_statistics_cp_samples_probes_placenta_cvs))
colnames(scaled_meQTLs_statistics_cp_samples_probes_placenta_cvs) <- gsub('.y', '.cvs', colnames(scaled_meQTLs_statistics_cp_samples_probes_placenta_cvs))
```


```{r, warning=F}
save(scaled_meQTLs_statistics_cp_samples_probes_placenta_cvs, file = here("02_Data/MatrixEQTL_Output", "scaled_meQTLs_statistics_cp_samples_probes_placenta_cvs.Rdata"))
```

### amount associaitons

```{r}
number_positive_placenta / total_number_tests_placenta
number_positive_cvs / total_number_tests_cvs
```



###################################################################################################################################################

# Placenta vs. CVS

```{r}
load(here("02_Data/MatrixEQTL_Output", "scaled_meQTLs_placenta_statistics_cp_samples_probes.Rdata"))
load(here("02_Data/MatrixEQTL_Output", "scaled_meQTLs_cvs_statistics_cp_samples_probes.Rdata"))

load(here("02_Data/MatrixEQTL_Output", "scaled_meQTLs_statistics_cp_samples_probes_placenta_cvs.Rdata"))
load(here("02_Data/CVS_Placenta/", "common_snp_cpg_placenta_cvs.Rdata"))
```

*common SNP-Gene pairs*
```{r}
dim(scaled_meQTLs_statistics_cp_samples_probes_placenta_cvs)

length(unique(scaled_meQTLs_statistics_cp_samples_probes_placenta_cvs$eSNPs))
length(unique(scaled_meQTLs_statistics_cp_samples_probes_placenta_cvs$eCpG))

dim(scaled_meQTLs_statistics_cp_samples_probes_placenta_cvs[scaled_meQTLs_statistics_cp_samples_probes_placenta_cvs$inboot.placenta == "yes", ])
dim(scaled_meQTLs_statistics_cp_samples_probes_placenta_cvs[scaled_meQTLs_statistics_cp_samples_probes_placenta_cvs$inboot.cvs == "yes", ])
```

```{r}
scaled_meQTLs_statistics_cp_samples_probes_placenta_cvs$abs.beta.placenta <- abs(scaled_meQTLs_statistics_cp_samples_probes_placenta_cvs$beta.placenta)
scaled_meQTLs_statistics_cp_samples_probes_placenta_cvs$abs.beta.cvs <- abs(scaled_meQTLs_statistics_cp_samples_probes_placenta_cvs$beta.cvs)
```


Select those previously found 19,583 common eQTMs

```{r}
common_pairs_previous_meqtls <- merge(scaled_meQTLs_statistics_cp_samples_probes_placenta_cvs, common_snp_cpg_placenta_cvs[,c("eSNPs", "eCpG")], by=c("eSNPs", "eCpG"))
nrow(common_pairs_previous_meqtls)
```
16,305 can be found


```{r}
# get absolute betas for placenta and cvs
common_pairs_previous_meqtls_absolute_betas <- common_pairs_previous_meqtls[ ,c("eSNPs", "eCpG", "abs.beta.placenta", "abs.beta.cvs")]
colnames(common_pairs_previous_meqtls_absolute_betas) <- c("eSNPs", "eCpG", "Placenta", "CVS")
common_pairs_previous_meqtls_absolute_betas$id_qtl <- rownames(common_pairs_previous_meqtls_absolute_betas)
```

```{r}
common_pairs_previous_meqtls_scaled_statistics_absolute_betas_long <- reformat_long_plot(data = common_pairs_previous_meqtls_absolute_betas)
```

```{r}
PlotQQ(common_pairs_previous_meqtls_absolute_betas$Placenta-common_pairs_previous_meqtls_absolute_betas$CVS)
gghistogram(common_pairs_previous_meqtls_absolute_betas$Placenta-common_pairs_previous_meqtls_absolute_betas$CVS)
```
```{r}
common_pairs_previous_meqtls_scaled_statistics_absolute_betas_long %>%
  group_by(tissue) %>%
  summarise_at(vars(absolute.beta), funs(mean,sd, median, IQR))
```

```{r}
t.test(common_pairs_previous_meqtls_absolute_betas$Placenta, common_pairs_previous_meqtls_absolute_betas$CVS, alternative = "greater", paired = T)
wilcox.test(common_pairs_previous_meqtls_absolute_betas$Placenta, common_pairs_previous_meqtls_absolute_betas$CVS, alternative = "greater", paired = T)
```

```{r}
# use rstatix to compute paired t test
t.test_common_meqtls <- common_pairs_previous_meqtls_scaled_statistics_absolute_betas_long  %>% 
  t_test(absolute.beta ~ tissue, paired = TRUE) %>%
  add_significance()

# Add p-value and significance levels
t.test_common_meqtls <- t.test_common_meqtls %>% add_xy_position(x = "tissue")

t.test_common_meqtls
```

check wilcoxon test, too
```{r}
common_pairs_previous_meqtls_scaled_statistics_absolute_betas_long  %>% 
  wilcox_test(absolute.beta ~ tissue, paired = TRUE) %>%
  add_significance()

common_pairs_previous_meqtls_scaled_statistics_absolute_betas_long  %>% 
  wilcox_effsize(absolute.beta ~ tissue, paired = TRUE)
```

```{r}
boxplot_common_hits_top_meqtls <- boxplot_paired_data(data = common_pairs_previous_meqtls_scaled_statistics_absolute_betas_long, textlabel = "paired T-Test, t(16,304) = 32.24, p < 0.001, n = 16,305")
boxplot_common_hits_top_meqtls$layers[[1]] <- NULL
boxplot_common_hits_top_meqtls$layers[[2]] <- NULL
boxplot_common_hits_top_meqtls
```


```{r}
save(common_pairs_previous_meqtls_scaled_statistics_absolute_betas_long, file = here("02_Data/CVS_Placenta/", "common_pairs_previous_meqtls_scaled_statistics_absolute_betas_long.Rdata"))
```








